### PR TITLE
Create resolver api

### PR DESCRIPTION
**Pull request by jbertel2**
I would like to point out a potential issue, hence creating a PR to discuss this

### CODE REVIEW

File: normalize.js [hunk 1]
```diff
@@ -332,41 +328,39 @@ var buildCustomSchema = exports.buildCustomSchema = function (schema, types, par
...
+          } else {
+            fields[field.uid] = '' + newparent;
           }
         }
+
         break;
       case 'blocks':
         var blockparent = parent.concat('_', field.uid);
-        var blockType = buildBlockCustomSchema(field.blocks, types, blockparent, prefix);
+        var blockType = buildBlockCustomSchema(field.blocks, types, references, blockparent, prefix);
```
Comment (jbertel2) [inline normalize.js:363]: why are the groups not being passed as a parameter here?

This leads to crashes during schema generation for our usecase

Comment (asmit-patil) [reply normalize.js]: Thanks for the notifying this. We have added the missing param.

Comment (jbertel2) [reply normalize.js]: i can confirm this works for us.

**Comment by jbertel2** [comment]
I still notice one case that doesn't work, probably related to schema generation.

When i query any union field for which the schema is generated by this branch, i get a graphql runtime error:

e.g. 

```
query MyQuery {
  contentstackMediaCenter {
    videos {
      related_to {
        ... on Contentstack_event {
          title
        }
        ... on Contentstack_specialty {
          title
        }
        ... on Contentstack_product {
          title
        }
      }
    }
  }
}
```

results in 

```
{
  "errors": [
    {
      "message": "Cannot read property 'type' of undefined",
      "locations": [
        {
          "line": 4,
          "column": 7
        }
      ],
      "path": [
        "contentstackMediaCenter",
        "videos",
        0,
        "related_to",
        0
      ],
      "stack": [
        "TypeError: Cannot read property 'type' of undefined",
        "    at /Users/jbertel2/Dev/jmc_frontend/node_modules/gatsby/src/schema/schema.js:406:57",
        "    at completeAbstractValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:652:21)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:585:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)",
        "    at /Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:618:25",
        "    at Array.forEach (<anonymous>)",
        "    at forEach (/Users/jbertel2/Dev/jmc_frontend/node_modules/iterall/index.js:83:25)",
        "    at completeListValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:614:24)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:573:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)",
        "    at resolveField (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:435:10)",
        "    at executeFields (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:275:18)",
        "    at collectAndExecuteSubfields (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:713:10)",
        "    at completeObjectValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:703:10)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:591:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)"
      ]
    },
    {
      "message": "Cannot read property 'type' of undefined",
      "locations": [
        {
          "line": 4,
          "column": 7
        }
      ],
      "path": [
        "contentstackMediaCenter",
        "videos",
        1,
        "related_to",
        0
      ],
      "stack": [
        "TypeError: Cannot read property 'type' of undefined",
        "    at /Users/jbertel2/Dev/jmc_frontend/node_modules/gatsby/src/schema/schema.js:406:57",
        "    at completeAbstractValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:652:21)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:585:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)",
        "    at /Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:618:25",
        "    at Array.forEach (<anonymous>)",
        "    at forEach (/Users/jbertel2/Dev/jmc_frontend/node_modules/iterall/index.js:83:25)",
        "    at completeListValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:614:24)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:573:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)",
        "    at resolveField (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:435:10)",
        "    at executeFields (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:275:18)",
        "    at collectAndExecuteSubfields (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:713:10)",
        "    at completeObjectValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:703:10)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:591:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)"
      ]
    },
    {
      "message": "Cannot read property 'type' of undefined",
      "locations": [
        {
          "line": 4,
          "column": 7
        }
      ],
      "path": [
        "contentstackMediaCenter",
        "videos",
        2,
        "related_to",
        0
      ],
      "stack": [
        "TypeError: Cannot read property 'type' of undefined",
        "    at /Users/jbertel2/Dev/jmc_frontend/node_modules/gatsby/src/schema/schema.js:406:57",
        "    at completeAbstractValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:652:21)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:585:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)",
        "    at /Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:618:25",
        "    at Array.forEach (<anonymous>)",
        "    at forEach (/Users/jbertel2/Dev/jmc_frontend/node_modules/iterall/index.js:83:25)",
        "    at completeListValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:614:24)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:573:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)",
        "    at resolveField (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:435:10)",
        "    at executeFields (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:275:18)",
        "    at collectAndExecuteSubfields (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:713:10)",
        "    at completeObjectValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:703:10)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:591:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)"
      ]
    },
    {
      "message": "Cannot read property 'type' of undefined",
      "locations": [
        {
          "line": 4,
          "column": 7
        }
      ],
      "path": [
        "contentstackMediaCenter",
        "videos",
        3,
        "related_to",
        0
      ],
      "stack": [
        "TypeError: Cannot read property 'type' of undefined",
        "    at /Users/jbertel2/Dev/jmc_frontend/node_modules/gatsby/src/schema/schema.js:406:57",
        "    at completeAbstractValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:652:21)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:585:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)",
        "    at /Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:618:25",
        "    at Array.forEach (<anonymous>)",
        "    at forEach (/Users/jbertel2/Dev/jmc_frontend/node_modules/iterall/index.js:83:25)",
        "    at completeListValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:614:24)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:573:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)",
        "    at resolveField (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:435:10)",
        "    at executeFields (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:275:18)",
        "    at collectAndExecuteSubfields (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:713:10)",
        "    at completeObjectValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:703:10)",
        "    at completeValue (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:591:12)",
        "    at completeValueCatchingError (/Users/jbertel2/Dev/jmc_frontend/node_modules/graphql/execution/execute.js:495:19)"
      ]
    }
  ],
  "data": {
    "contentstackMediaCenter": {
      "videos": [
        {
          "related_to": [
            null
          ]
        },
        {
          "related_to": [
            null
          ]
        },
        {
          "related_to": [
            null
          ]
        },
        {
          "related_to": [
            null
          ]
        }
      ]
    }
  }
}
```

**Comment by jbertel2** [comment]
Screenshot 2020-08-12 at 21 01 25
As can be seen in the above screenshot, some nodes don't have the expected internal field set...

File: normalize.js [hunk 2]
```diff
@@ -380,10 +374,49 @@ var buildCustomSchema = exports.buildCustomSchema = function (schema, types, par
...
+            unions.push(referenceType);
+            var type = 'type ' + referenceType + ' implements Node { title: String! }';
+            types.push(type);
+          });
+          var name = '';
+          name = name.concat(unions.join(''), '_Union');
+          unionType = unionType.concat('_Union = ', unions.join(' | '));
+          types.push(unionType);
+
+          references.push({
```
Comment (jbertel2) [inline normalize.js:402]: i believe just pushing an object with just a parent and a uid is not sufficient to resolve queries afterwards, as demonstrated by the runtime error that is looking for an internal field with a type

Comment (jbertel2) [reply normalize.js:402]: while debugging, i noticed the related_to field being resolved by the default resolver and not the custom resolver we register...

Comment (jbertel2) [reply normalize.js:402]: it looks like not all reference fields get our custom resolvers registered, but some fallback to the default resolver, with the runtime error as consequence. Not sure why yet...

File: gatsby-node.js [hunk 3]
```diff
@@ -253,4 +260,39 @@ exports.sourceNodes = function () {
...
+            });
+          });
+          return nodesData;
+        }
+        return [];
+      }
+    });
+  });
+  groups.forEach(function (group) {
+    resolvers[group.parent] = (0, _defineProperty3.default)({}, group.field.uid, {
```
Comment (jbertel2) [inline gatsby-node.js:288]: I believe this is the root cause of the remaining problem. References and groups can have the same parent, or multiple references/groups can have the same parent. In these cases, we are overwriting the previous resolver...

Comment (jbertel2) [reply gatsby-node.js:288]: See https://github.com/contentstack/gatsby-source-contentstack/pull/43/commits/a7354b414550346a379817b29a25fc749061c29f for a potential fix. This worked in my case...

Comment (asmit-patil) [reply gatsby-node.js:288]: Thanks @jbertel2 for notifying this issue and fixing it. We appreciate your contribution here.

**Pull request merged**